src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs

```diff
@@ -0,0 +1,232 @@
+            this._jsonRpc.StartListening();
+
+            this._diagnosticService = diagnosticService;
+            this._diagnosticService.DiagnosticsUpdated += DiagnosticService_DiagnosticsUpdated;
+        }
+
+        /// <summary>
+        /// Handle the LSP initialize request by storing the client capabilities
+        /// and responding with the server capabilities.
+        /// The specification assures that the initialize request is sent only once.
+        /// </summary>
+        [JsonRpcMethod(Methods.InitializeName)]
+        public InitializeResult Initialize(JToken input)
+        {
+            // InitializeParams only references ClientCapabilities, but the VS LSP client
+            // sends additional VS specific capabilities, so directly deserialize them into the VSClientCapabilities
+            // to avoid losing them.
+            this._clientCapabilities = input["capabilities"].ToObject<VSClientCapabilities>();
+
+            return new InitializeResult
+            {
+                Capabilities = new VSServerCapabilities
+                {
+                    DocumentHighlightProvider = true,
+                    DocumentSymbolProvider = true,
+                    DocumentFormattingProvider = true,
+                    DocumentRangeFormattingProvider = true,
+                    DocumentOnTypeFormattingProvider = new DocumentOnTypeFormattingOptions { FirstTriggerCharacter = "}", MoreTriggerCharacter = new[] { ";", "\n" } },
+                    DefinitionProvider = true,
+                    CompletionProvider = new CompletionOptions { ResolveProvider = true, TriggerCharacters = new[] { "." } },
+                    ImplementationProvider = true,
+                    ReferencesProvider = true,
+                    SignatureHelpProvider = new SignatureHelpOptions { TriggerCharacters = new[] { "(", "," } },
+                    WorkspaceSymbolProvider = true,
+                }
+            };
+        }
+
+        [JsonRpcMethod(Methods.InitializedName)]
+        public Task Initialized() => Task.CompletedTask;
+
+        [JsonRpcMethod(Methods.ShutdownName)]
+        public object? Shutdown(CancellationToken _) => null;
+
+        [JsonRpcMethod(Methods.ExitName)]
+        public void Exit()
+        {
+        }
+
+        /// <summary>
+        /// Go to definition calls to get third party definitions (xaml) and to get those definitions it expectes to be called on the UI thread.
+        /// <see cref="VisualStudioSymbolNavigationService.WouldNotifyToSpecificSymbol"/>
+        /// </summary>
+        [JsonRpcMethod(Methods.TextDocumentDefinitionName)]
+        public async Task<object> GetTextDocumentDefinitionAsync(JToken input, CancellationToken cancellationToken)
+        {
+            var textDocumentPositionParams = input.ToObject<TextDocumentPositionParams>();
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
```

Comment:
dumb question.  why is this here, instead of inside `._protocol.GoToDefinitionAsync` when it needs to call out to 3rd parties?

Reply:
That code lives in the features layer, so I was attempting to keep it 'agnostic' as much as possible of UI concerns like this and move all the required UI switching up to the top (so we can easily keep track of it and delete it).

Reply:
i don't see how htat could possibly work.  once you call GoToDefinitionAsync then the *impl* of that might do `.ConfiureAwait(false)` *inside* of it.  Which would then cause the same failures you're worried about when results are reported and need to be broadcast on the UI thread.

Reply:
Ultimately, only teh code that is actually doing the broadcasting is the one that can ensure that it is on the right thread.  The caller (you) cannot do this with async methods as there's nothing to ensure this UI thread that you're now on has any bearing later on after otehr async work is done.

Reply:
makes sense.

Will just move the whole implementation up to the VS layer and do the switch there.  once the UI dependency is removed I'll move it back down.

Reply:
So two followups I learned after more digging.

1.  Goto definition doesn't actually need the UI thread.
It calls into [FindDefinitionsAsync](http://sourceroslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/GoToDefinition/AbstractGoToDefinitionService.cs,495ff7831b9e7b4d,references) to get the defintion which doesn't seem to require the UI thread (correct me if I'm wrong).  It also doesn't seem to get third party references either (seems like missing support, will file an issue to track that).

2.  I also learned it isn't safe to use JTF to switch to the main thread across a stream json rpc boundary.  It could still deadlock as the call isn't tracked across rpc.  So for FAR, I will be unable to migrate to the IlanguageClient until we remove the UI thread dependency.

thoughts @CyrusNajmabadi @jasonmalinowski ?

Reply:
We bite the bullet and say that if someone wants to participate in FAR they need to use a new free-threaded async method that uses modern practices.

Reply:
(that's just my opinion of course).  But having these ancient sta interfaces (from literally 15+ years ago) hold us back is silly.  Let's add the new interface.  If the scenario matters to existing clients they can implement this interface.

Reply:
moved back to the liveshare API for now - no concerns using jtf there.  Have https://github.com/dotnet/roslyn/issues/38477 for fixing this so we can properly fix it.